File: src/ros2_bridge/mappers/foxglove_msgs/pose.rs
//! Typed mapper for `foxglove_msgs/msg/Pose`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn pose_to_bus(msg: ros_env::foxglove_msgs::msg::Pose) -> crate::foxglove_msgs::msg::v1::Pose {
    crate::foxglove_msgs::msg::v1::Pose {
        position: Some(crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_bus(msg.position)),
        orientation: Some(crate::ros2_bridge::mappers::foxglove_msgs::quaternion::quaternion_to_bus(msg.orientation)),
    }
}

pub(crate) fn pose_to_ros(bus: crate::foxglove_msgs::msg::v1::Pose) -> ros_env::foxglove_msgs::msg::Pose {
    ros_env::foxglove_msgs::msg::Pose {
        position: crate::ros2_bridge::mappers::foxglove_msgs::vector3::vector3_to_ros(bus.position.unwrap_or_default()),
        orientation: crate::ros2_bridge::mappers::foxglove_msgs::quaternion::quaternion_to_ros(bus.orientation.unwrap_or_default()),
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct FoxgloveMsgsPoseMapper;

impl TypedTopicMapper for FoxgloveMsgsPoseMapper {
    type Ros = ros_env::foxglove_msgs::msg::Pose;
    type Bus = crate::foxglove_msgs::msg::v1::Pose;

    fn type_name(&self) -> &'static str {
        "foxglove_msgs/msg/Pose"
    }

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(pose_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(pose_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/foxglove_msgs/pose.py
"""Generated mapper for `foxglove_msgs/msg/Pose`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.foxglove_msgs.vector3 import vector3_to_bus, vector3_to_ros
from robot_bus.ros2_bridge.mappers.foxglove_msgs.quaternion import quaternion_to_bus, quaternion_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.foxglove_msgs.msg.v1 import Pose as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def pose_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.position.CopyFrom(vector3_to_bus(msg.position))
    bus.orientation.CopyFrom(quaternion_to_bus(msg.orientation))
    return bus


def pose_to_ros(bus):
    from foxglove_msgs.msg import Pose as RosMsg

    out = RosMsg()
    out.position = vector3_to_ros(bus.position)
    out.orientation = quaternion_to_ros(bus.orientation)
    return out


class FoxgloveMsgsPoseMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from foxglove_msgs.msg import Pose as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return pose_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return pose_to_ros(bus)
